octopusdeploy_framework/datasource_tenant_projects.go

```diff
@@ -0,0 +1,201 @@
+							Description: "The environment IDs associated with this tenant.",
+							ElementType: types.StringType,
+							Computed:    true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func (t *tenantProjectsDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var data TenantProjectsDataModel
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	tenantIDs := util.ExpandStringList(data.TenantIDs)
+	projectIDs := util.ExpandStringList(data.ProjectIDs)
+	environmentIDs := util.ExpandStringList(data.EnvironmentIDs)
+	spaceID := data.SpaceID.ValueString()
+
+	if tenantIDs == nil && projectIDs == nil {
+		resp.Diagnostics.AddError("must provide at least one tenant or one project", "tenant IDs and project IDs are nil")
+	}
+
+	var tenantData []*tenants.Tenant
+	if tenantIDs == nil {
+		tenantData = make([]*tenants.Tenant, 0)
+		for _, projectID := range projectIDs {
+			// todo: should consider using concurrency
+			tenantsByProjectID, err := getTenantByProjectID(t.Client, projectID, spaceID)
+			if err != nil {
+				resp.Diagnostics.AddError("unable to load tenant data for project", err.Error())
+				return
+			}
+			tenantData = append(tenantData, tenantsByProjectID...)
+		}
+	} else {
+		tenantQuery := tenants.TenantsQuery{
+			IDs: tenantIDs,
+		}
+		tenantResource, err := tenants.Get(t.Client, spaceID, tenantQuery)
+		if err != nil {
+			resp.Diagnostics.AddError("unable to load tenant data", err.Error())
+			return
+		}
+		tenantData, err = tenantResource.GetAllPages(t.Client.Sling())
+		if err != nil {
+			resp.Diagnostics.AddError("unable to load tenant data", err.Error())
+			return
+		}
+	}
+
+	// todo: refactor
+	flattenedTenantProjects := make([]any, 0)
+	for _, tenant := range tenantData {
+		if projectIDs == nil {
+			for projectID, envIDs := range tenant.ProjectEnvironments {
+				if environmentIDs != nil {
+					for _, envID := range envIDs {
+						if slices.Contains(environmentIDs, envID) {
+							flattenedTenantProjects = append(flattenedTenantProjects, flattenTenantProject(tenant, projectID))
+							break
+						}
+					}
+				} else {
+					flattenedTenantProjects = append(flattenedTenantProjects, flattenTenantProject(tenant, projectID))
+				}
+			}
+			continue
+		}
+		for _, projectID := range projectIDs {
+			if envIDs, ok := tenant.ProjectEnvironments[projectID]; ok {
+				if environmentIDs != nil {
+					for _, envID := range envIDs {
+						if slices.Contains(environmentIDs, envID) {
+							flattenedTenantProjects = append(flattenedTenantProjects, flattenTenantProject(tenant, projectID))
+							break
+						}
+					}
+				} else {
+					flattenedTenantProjects = append(flattenedTenantProjects, flattenTenantProject(tenant, projectID))
+				}
+			}
+
+		}
+	}
+
+	data.TenantProjects, _ = types.ListValueFrom(ctx, types.ObjectType{AttrTypes: tenantProjectType()}, flattenedTenantProjects)
```

Comment:
Nit: If we're not handling the error output here you can use types.ListValueMust and return that directly.

Reply:
thanks for the tip!